miriway.cpp

```diff
@@ -94,9 +95,25 @@ struct ShellComponents
             launch(command);
         }
     }
+
+    bool try_add_lockscreen_app(std::vector<std::string> const& cmd)
+    {
+        if (lockscreen_app.has_value())
+            return false;
+
+        lockscreen_app = cmd;
+        return true;
+    }
+
+    void launch_lockscreen() const
+    {
+        if (lockscreen_app.has_value())
+            launch(lockscreen_app.value());
+    }
```

Comment:
I can't help feeling there's no reason to insert this logic into `ShellComponents`.

A `LockScreen` class could own everything: the option and result, the `SessionLockListener`, the registration with ShellPids, adding the Wayland extension. Something like this...

```c++
class  LockScreen
{
public:
    explicit LockScreen(std::function<void>(std::string name) conditionally_enable)
    (
        conditionally_enable(WaylandExtensions::ext_session_lock_manager_v1);
    }

    void operator()(mir::Server& server) const
    {
        lockscreen_option(server);
        session_locker(server);
    }

private:
    std::optional<std::string> lockscreen_app;
    
    ConfigurationOption const lockscreen_option{
        [this](std::string const& app) {
            lockscreen_app = app;
        }

        "lockscreen-app",
        "Lockscreen app to be triggered when the compositor session is locked (may only be specified once)"
    };

    SessionLockListener const session_locker(
            [this] { if (lockscreen_app) shell_launch( split_command(*lockscreen_app)); },
            [] {});
};
```

Reply:
Implemented. Feels a lot better